x-pack/plugins/monitoring/server/telemetry_collection/get_logstash_stats.ts

```diff
@@ -0,0 +1,423 @@
+      const vertices = thisLogstashStatePipeline.representation?.graph?.vertices;
+
+      if (vertices !== undefined) {
+        vertices.forEach((vertex) => {
+          const configName = vertex.config_name;
+          const pluginType = vertex.plugin_type;
+          let pipelineConfig = vertex.meta?.source?.protocol;
+
+          if (pipelineConfig !== undefined) {
+            if (pipelineConfig === 'string' || pipelineConfig === 'str') {
+              pipelineConfig = 'string';
+            } else if (pipelineConfig === 'x-pack-config-management') {
+              pipelineConfig = 'xpack';
+            } else {
+              pipelineConfig = 'file';
+            }
+            if (!pipelineStats.hasOwnProperty('sources')) {
+              pipelineStats.sources = {};
+            }
+            pipelineStats.sources![pipelineConfig] = true;
+          }
+          if (configName !== undefined && pluginType !== undefined) {
+            incrementByKey(plugins[clusterUuid], `logstash-${pluginType}-${configName}`);
+          }
+        });
+      }
+    }
+  });
+  if (clusterStats !== undefined) {
+    clusterStats.plugins = mapToList(plugins[clusterUuid], 'name');
+  }
+}
+
+export async function fetchLogstashStats(
+  callCluster: LegacyAPICaller,
+  clusterUuids: string[],
+  { page = 0, ...options }: { page?: number } & LogstashProcessOptions
+): Promise<void> {
+  const params = {
+    headers: {
+      'X-QUERY-SOURCE': TELEMETRY_QUERY_SOURCE,
+    },
+    index: INDEX_PATTERN_LOGSTASH,
+    ignoreUnavailable: true,
+    filterPath: [
+      'hits.hits._source.cluster_uuid',
+      'hits.hits._source.type',
+      'hits.hits._source.source_node',
+      'hits.hits._source.agent.type',
+      'hits.hits._source.logstash_stats.pipelines.id',
+      'hits.hits._source.logstash_stats.pipelines.ephemeral_id',
+      'hits.hits._source.logstash_stats.pipelines.queue.type',
+      'hits.hits._source.logstash_stats.logstash.version',
+      'hits.hits._source.logstash_stats.logstash.uuid',
+    ],
+    body: {
+      query: createQuery({
+        filters: [
+          { terms: { cluster_uuid: clusterUuids } },
+          {
+            bool: {
+              must: { term: { type: 'logstash_stats' } },
+            },
+          },
+        ],
+      }),
+      from: page * HITS_SIZE,
+      collapse: { field: 'logstash_stats.logstash.uuid' },
+      sort: [{ ['logstash_stats.timestamp']: { order: 'desc', unmapped_type: 'long' } }],
+      size: HITS_SIZE,
+    },
+  };
+
+  const results = await callCluster<SearchResponse<LogstashStats>>('search', params);
+  const hitsLength = results?.hits?.hits.length || 0;
+
+  if (hitsLength > 0) {
+    // further augment the clusters object with more stats
+    processStatsResults(results, options);
+
+    if (hitsLength === HITS_SIZE) {
+      // call recursively
+      const nextOptions = {
+        page: page + 1,
+        ...options,
+      };
```

Comment:
Mentioned in another comment. But unfortunately, this type of pagination is not supported by ES because the max_window (from + size) can't exceed 10k docs by default (unless `.monitoring` indices have any other config).

Reply:
As @robbavey pointed out, this is following the same approach as the [get_beats_stats](https://github.com/elastic/kibana/blob/master/x-pack/plugins/monitoring/server/telemetry_collection/get_beats_stats.ts#L370) piece of logic.

@chrisronline, do you know if `.monitoring-*` indices have a custom `index.max_result_window` setting that allows pagination for more than 10k docs? Or if it's even possible that we reach 10k docs on these types of queries?

Reply:
I guess the key point lies here:
```
      collapse: { field: 'logstash_stats.logstash.uuid' },
...
      collapse: { field: 'logstash_state.pipeline.ephemeral_id' },
```
Is it possible that we have more than 10k `logstash.uuid` or `ephemeral_id`s? If it's unlikely, we could merge this PR and create an enhancement to implement pagination with a safer mechanism that can be used by both: Beats and Logstash.

What do you think?